src/hiccup/compiler.clj

```diff
@@ -73,37 +73,48 @@
   (or content
       (and (html-mode?) (not (void-tags tag)))))
 
-(defn- merge-classes [class classes]
+(defn- merge-classes [class classes compiling?]
   (cond
-    (nil? class)    classes
-    (string? class) (str classes " " class)
-    :else           (str classes " " (str/join " " class))))
-
-(defn- merge-attributes [map-attrs id classes]
+    (nil? class)                  classes
+    (string? class)               (str classes " " class)
+    (and compiling?
+         (symbol? class))         (list `str classes " " class)
+    (and (not compiling?)
+         (instance? Named class)) (str classes " " (name class))
+    :else                         (str classes " " (str/join " " class))))
+
+(defn- merge-attributes [map-attrs id classes compiling?]
   (-> map-attrs
       (cond-> id      (assoc :id (or (:id map-attrs) id)))
-      (cond-> classes (assoc :class (merge-classes (:class map-attrs) classes)))))
+      (cond-> classes (assoc :class (merge-classes (:class map-attrs) classes compiling?)))))
 
-(defn normalize-element
-  "Ensure an element vector is of the form [tag-name attrs content]."
-  [[tag & content]]
+(defn- normalize-element-with-compile-mode
+  "Ensure an element vector is of the form [tag-name attrs content], choosing
+   a different class merge strategy fi rendering or compiling."
+  [[tag & content] compiling?]
   (when (not (or (keyword? tag) (symbol? tag) (string? tag)))
     (throw (IllegalArgumentException. (str tag " is not a valid element name."))))
   (let [[_ tag id class] (re-matches re-tag (util/as-str tag))
         classes          (if class (str/replace class "." " "))
         map-attrs        (first content)]
     (if (map? map-attrs)
-      [tag (merge-attributes map-attrs id classes) (next content)]
+      [tag (merge-attributes map-attrs id classes compiling?) (next content)]
       [tag {:id id, :class classes} content])))
 
+(defn normalize-element
```

Comment:
We need to keep `normalize-element` behaving the same, so normalizing elements in a non-compiled context. We can't change its functionality and break compatibility.